src/meta/src/hummock/manager/compaction.rs

```diff
@@ -117,76 +121,186 @@ impl HummockManager {
+                        {
+                            tracing::warn!(
+                                error = %e.as_report(),
+                                "Failed to send task {} to {}",
+                                task_id,
+                                compactor.context_id(),
+                            );
+                            // failed_tasks.push(task_id);
+                            send_fail = true;
+                            break;
+                        }
+                    }
+
+                    existed_groups.retain(|group_id| !no_task_groups.contains(group_id));
+
+                    // cancel all tasks
+                    if send_fail {
+                        failed_tasks.extend(compact_tasks.iter().map(|t| t.task_id));
+                    }
+
+                    (failed_tasks, generated_task_count)
+                }
+                Err(err) => {
+                    tracing::warn!(error = %err.as_report(), "Failed to get compaction task");
+                    (failed_tasks, generated_task_count)
+                }
+            }
+        }
+
+        async fn handle_pick_tasks(
+            hummock_manager: &HummockManager,
+            pull_task_count: usize,
+            groups: Vec<u64>,
+            compactor: Arc<Compactor>,
+            selector: &mut Box<dyn CompactionSelector>,
+            no_task_groups: &mut HashSet<CompactionGroupId>,
+            existed_groups: &mut Vec<CompactionGroupId>,
+        ) -> (Vec<u64>, usize) {
+            let failed_tasks = vec![];
+            let mut generated_task_count = 0;
+            while generated_task_count < pull_task_count {
+                let compact_ret = hummock_manager
+                    .get_compact_tasks(
+                        groups.clone(),
+                        pull_task_count - generated_task_count,
+                        selector,
+                    )
+                    .await;
+                let (failed_tasks, picked_tasks_count) = tasks_result_handler(
+                    compact_ret,
+                    compactor.clone(),
+                    no_task_groups,
+                    existed_groups,
+                );
+
+                // if no task or failed task, break
+                if picked_tasks_count == 0 || !failed_tasks.is_empty() {
+                    break;
+                }
+
+                generated_task_count += picked_tasks_count;
+            }
+
+            (failed_tasks, generated_task_count)
+        }
+
         if let Some(compactor) = self.compactor_manager.get_compactor(context_id) {
             let (groups, task_type) = self.auto_pick_compaction_groups_and_type().await;
+
+            let mut generated_task_count = 0;
+            let mut existed_groups = groups.clone();
+            let mut no_task_groups: HashSet<CompactionGroupId> = HashSet::default();
+            let mut failed_tasks = vec![];
```

Comment:
nits: there are two patterns to mutate these variables
- `existed_groups` and `no_task_groups` are mutabled inside `handle_pick_tasks` by passing `&mut`
- `generated_task_count` and `failed_tasks` are mutabled outside of `handle_pick_tasks` by using the returned result.

Can we converge them to one pattern?